aten/src/ATen/native/cpu/AdaptiveAvgPoolKernel.cpp

```diff
@@ -149,6 +149,105 @@ void cpu_adaptive_avg_pool_channels_last(
   }
 }
 
+template <>
+void cpu_adaptive_avg_pool_channels_last<BFloat16>(
```

Comment:
Can you please clarify why we need separate kernel code for channels last bfloat16 but can go with same kernel in contiguous

Reply:
The adaptive {avg|max} pooling kernel is involving reduction.
For `bfloat16` input we should be using `float32` as accumulate type.

On 'contiguous' memory formats, it is purely scalar logic, we don't have a chance to vectorize it so directly casting dtype is proper. But since we can not vectorize the dtype conversion logic, **scalar bfloat16 is going to be slower than original float32** kernel (on most occasions).

On 'channels last' memory formats, we can vectorize the kernel along 'channel' dimension. And for the bfloat16 implementation, we need to allocate a n immediate buffer (float32) to store the 'sum'. Thus the reduction is done on float32 (reduce rounding error and improve performance). I thought mixing bfloat16 kernel with original float32 is going to be a little messy so i wrote a separate <BFLoat16> kernel. And **bfloat16 kernel is faster than float32 when this is a) vectorized kernel; b) memory bandwidth bound kernel**